src/Wave.hpp

```diff
@@ -0,0 +1,114 @@
+#ifndef GAZEBO_WAVE_PLUGIN_HPP
+#define GAZEBO_WAVE_PLUGIN_HPP
+
+#include <gazebo/common/common.hh>
+#include <gazebo/physics/physics.hh>
+#include <gazebo/transport/transport.hh>
+#include <ignition/math/Vector3.hh>
+#include "base/Time.hpp"
+
+namespace gazebo_usv
+{
+    class Wave
+    {
+        typedef gazebo::physics::ModelPtr ModelPtr;
+        typedef gazebo::physics::LinkPtr LinkPtr;
+        typedef gazebo::transport::NodePtr NodePtr;
+        typedef gazebo::transport::SubscriberPtr SubscriberPtr;
+
+    public:
+        // Parameters used to calculate the Wave effects 
+        struct EffectParameters
+        {
+            double torque_constant;
+        };
+        // Wave force and torque
+        struct Effects
+        {
+            ignition::math::Vector3d force;
+            ignition::math::Vector3d torque;
+        };
+        //Wave phases to be defined
+        double phase_x;
+        double phase_y;
+        double phase_z;
+
+        Wave() = default;
+        /**
+         * @brief Construct a new Wave object.
+         * 
+         * PS: This constructor is used mainly for easier testing.
+         * 
+         * @param parameters 
+         */
+        Wave(EffectParameters const parameters);
+        ~Wave();
+
+        /**
+         * @brief Loads the model properties from the SDF file and the gazebo communication node.
+         * 
+         * @param _model model pointer
+         * @param _node gazebo node
+         * @param _sdf sdf element
+         */
+        void load(ModelPtr const _model, NodePtr const _node, sdf::ElementPtr const _sdf);
+
+        /**
+         * @brief Update the wave effects on the model.
+         * 
+         */
+        void update();
+
+        /**
+         * @brief Computes the wave effects on a vessel. 
+         * 
+         * Reference: 
+         *  - Fossen's Handbook of Marine Craft Hydrodynamics and Motion Control: pages 188 to 192.
+         * 
+         * PS: This method is public for testing purposes and shouldn't be used for other reasons.
+         * @param body2world_orientation vessel orientation in world frame 
+         * @param vessel_linear_vel_world vessel linear amplitude in world frame 
+         * @param wave_amplitude_world wave amplitude in world frame 
+         * @return Effects resulting force and torque to be applied at the vessel CoG. 
+         */
+        Effects computeEffects(double seconds, ignition::math::Vector3d const wave_amplitude_world, ignition::math::Vector3d const wave_frequency_world) const;
+
+    private:
+        ModelPtr mModel;
+        NodePtr mNode;
+        LinkPtr mLink;
+
+        SubscriberPtr mWaveAmplitudeSubscriber;
+        SubscriberPtr mWaveFrequencySubscriber;
+
+        EffectParameters mParameters;
+        ignition::math::Vector3d mWaveAmplitude{};
+        ignition::math::Vector3d mWaveFrequency{};
```

Comment:
Dont forget to rename the variable to account for the most recent naming guidelines (mCamelCase to m_snake_case for private member variables)